piwik for results opend in the list using ajax

jochen-lienhard opened the pull request:
In the current version there are two problems:

1) opening a title using accordion/tabs instead of view=full in searches.ini is not tracked.
2) switching the accordion/tabs in an opened title does not track the custom variables.

demiankatz commented:
@jochen-lienhard, it looks like you have changed the ajaxtabAction, which is used for dynamically loading the tabs on the record view. Without referencing the code, I can't recall if this is also used for accordions or if that is being loaded through some other route. In any case, I'm wondering if there are record-tracking implications for making this change. @EreMaijala, do you recall how we wanted to deal with record tabs in the past? Do we need to have different considerations for AJAX-loaded records in the search results (for Jochen's situation) vs. AJAX-loaded tabs in the standard record page (for the more common case)?

Apologies for not remembering all of this better -- just trying to help facilitate the right conversation to get this sorted out!

In any case, I do support the idea of having different layout templates for different contexts -- the trick is identifying all of the cases and making sure we do the right thing in each case.

EreMaijala commented:
I wrote about it on the mailing list, but anyway, it's down to semantics. In a way clicking the title to expand the record counts as a record view so it should be considered equal to loading the record view or switching tabs in it. It's kinda the same function with just a slightly different UI. Then again, it's not quite the same, and if there's a need to e.g. analyze usage patterns it would be very useful to track them as different actions. My own feeling is that tracking them separately would be preferred.

jochen-lienhard commented:
We can keep the ajaxtabAction as it is. If someone wish to track the custom variables, he can create an own lightbox.pthml in his theme without the lightbox = true call of piwik. Perhaps we can add a comment in the wiki (https://vufind.org/wiki/configuration:usage_stats) that describes this.

For the missing tracking of opening a title, I have another idea, that maybe solve the problem and make the ajax-Accordion-View template smaller. I will implement it an add it to this PR.

jochen-lienhard commented:
I think, my idea has to many side effects. The idea is to implement the core as a tab too, but this will effect the standard view too. 
Can we add something to the embedded_record.js so that the opening of a title can be tracked?

demiankatz commented:
If memory serves, when you click a title for the first time in the list, it makes some kind of AJAX call to load the whole tab or accordion mechanism... then the getAjaxTab routine is used to fill in the secondary tabs. Thus, this initial load should be trackable somehow or other, since that is what actually triggers the loading of the core data. Is that any help? Let me know if you need me to find the technical details and I can do so when I have a little more free time.

jochen-lienhard commented:
I don't know, why the inital load is not tracked. Can you ask Chris, if he has an idea, because he modified the embedded_record.js.

demiankatz commented:
Chris is on vacation at the moment, but I may be able to help.

Here is where the AJAX call is made to load the initial record data -- it uses the getRecordDetails AJAX method:

https://github.com/vufind-org/vufind/blob/master/themes/bootstrap3/js/embedded_record.js#L130

That AJAX method is now implemented here:

https://github.com/vufind-org/vufind/blob/master/module/VuFind/src/VuFind/AjaxHandler/GetRecordDetails.php

It operates by rendering one of the ajaxview templates:

https://github.com/vufind-org/vufind/blob/master/themes/bootstrap3/templates/record/ajaxview-tabs.phtml

or

https://github.com/vufind-org/vufind/blob/master/themes/bootstrap3/templates/record/ajaxview-accordion.phtml

Is it possible that we just need to add Piwik/Google Analytics logic to these two files, or else force the AJAX handler to render them in the context of the lightbox layout?

I'm always a little hazy on how Javascript gets executed when it's dynamically loaded as part of an AJAX process, so maybe it's not that simple... but it might be a place to start. Perhaps @EreMaijala can comment on whether or not this is workable.

jochen-lienhard commented:
I added the piwik call to the ajaxview templates. Now the first open is tracked. 

@demiankatz: Shall I remove the modification in getAjaxTab and do you add some comments about custom variables to the wiki?

demiankatz commented:
@jochen-lienhard, great, I'm glad that this solved the "first open" problem. I imagine we should probably add the Google Analytics hook here as well for consistency.

If things are working acceptably for you without the need to introduce the inline layout, we might as well revert that to prevent confusion. If you feel that additional notes/documentation are needed, they can be added to this wiki page:

https://vufind.org/wiki/configuration:usage_stats

We should probably add the URL of this page to the comments in config.ini as well so people know that it exists, especially if we add important notes there.

Please let me know if you need me to help with any of this -- but you have a better understanding of the needs, so I prefer not to take the lead on the documentation. :-)

Thanks!

jochen-lienhard commented:
I added the entry for google analytics and removed the inline template.

I have no access to the wiki, so I post here some text, that may be included.

Some Tips for Using Piwik

The custom variables are not tracked, when the content is loaded using ajax. If you wish to track the custom variables in this case too, remove the "lightbox = true" entry in the lightbox.phtml in the piwik call.

themes/bootstrap3/templates/layout/lightbox.phtml
Replace (line 2):
<?=$this->piwik(['lightbox' => true])?>
with:
<?=$this->piwik()?>

demiankatz commented:
Thanks, @jochen-lienhard, this looks good, and I have updated the wiki accordingly. Please let me know if you have any further problems with this situation!

jochen-lienhard commented:
At the moment I have no further problems, this PR can be closed.
